aten/src/TH/generic/THTensorRandom.c

```diff
@@ -111,18 +135,58 @@ void THTensor_(exponential)(THTensor *self, THGenerator *_generator, double lamb
 void THTensor_(standard_gamma)(THTensor *self, THGenerator *gen, THTensor *alpha)
 {
   THTensor_(resizeAs)(self, alpha);
-  TH_TENSOR_APPLY2(real, self, real, alpha, *self_data = THRandom_standard_gamma(gen, *alpha_data););
+  TH_TENSOR_APPLY2(real, self, real, alpha, {
+    double unused;
+    *self_data = THRandom_standard_gamma_fwd(gen, *alpha_data, &unused, &unused);
+  });
 }
 
-void THTensor_(standard_gamma_with_grad)(THTensor *self, THGenerator *gen, THTensor *alpha, THTensor *grad)
+void THTensor_(standard_gamma_fwd)(THTensor *self, THGenerator *gen, THTensor *alpha, THTensor *saved_u, THTensor *saved_x)
 {
+  alpha = THTensor_(newContiguous)(alpha);
   THTensor_(resizeAs)(self, alpha);
-  THTensor_(resizeAs)(grad, self);
-  TH_TENSOR_APPLY3(real, self, real, alpha, real, grad, {
-    double temp;
-    *self_data = (real)THRandom_standard_gamma_with_grad(gen, *alpha_data, &temp);
-    *grad_data = (real)temp;
-  });
+  THTensor* sample = THTensor_(newContiguous)(self);
+  TH_CHECK_CONTIGUOUS(saved_u);
+  TH_CHECK_CONTIGUOUS(saved_x);
+  TH_CHECK_SAME_SIZE(saved_u, alpha);
+  TH_CHECK_SAME_SIZE(saved_x, alpha);
+
+  real*const sample_data = THTensor_(data)(sample);
+  real*const alpha_data = THTensor_(data)(alpha);
+  real*const saved_u_data = THTensor_(data)(saved_u);
+  real*const saved_x_data = THTensor_(data)(saved_x);
+  for(int64_t i = 0, numel = THTensor_(nElement)(alpha); i < numel; ++i) {
```

Comment:
Any reason why you don't use OpenMP here?

Reply:
It's difficult to parallelize because `THGenerator *gen` is stateful. However this `_fwd()` step is much cheaper than the `_bwd()` step.